Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java

```diff
@@ -0,0 +1,701 @@
+		}
+		return sumArea;
+	}
+
+	static double[] getSurfaceVolume(final Collection<List<Point3f>> surfacePoints) {
+		return surfacePoints.parallelStream().mapToDouble(p -> {
+			if (p == null) {
+				return 0;
+			}
+			final CustomTriangleMesh mesh = new CustomTriangleMesh(p);
+			return Math.abs(mesh.getVolume());
+		}).toArray();
+	}
+
+	/**
+	 * Get the Feret diameter of a surface. Uses an inefficient brute-force
+	 * algorithm.
+	 *
+	 * @param surfacePoints points of all the particles.
+	 * @return Feret diameters of the surfaces.
+	 */
+	static double[] getFerets(final List<List<Point3f>> surfacePoints) {
+		Thread[] threads = Multithreader.newThreads();
+		final int nSurfaces = surfacePoints.size();
+		final double[] ferets = new double[nSurfaces];
+		AtomicInteger ai = new AtomicInteger(0);
+		for (int thread = 0; thread < threads.length; thread++) {
+			threads[thread] = new Thread(() -> {
+				for (int i = ai.getAndIncrement(); i < nSurfaces; i = ai.getAndIncrement()) {
+					final List<Point3f> surface = surfacePoints.get(i);
+				
+					if (surface == null) {
+						ferets[i] = Double.NaN;
+						continue;
+					}
+					
+					// check all the point pairs in this surface
+					ListIterator<Point3f> ita = surface.listIterator();
+					ListIterator<Point3f> itb;
+					Point3f a;
+					Point3f b;
+					double feret = 0;
+					while (ita.hasNext()) {
+						// for all the points
+						a = ita.next();
+						// check all the pairs after this point
+						// (the other direction has already been checked)
+						itb = surface.listIterator(ita.nextIndex());
+						while (itb.hasNext()) {
+							b = itb.next();
+							feret = Math.max(feret, a.distance(b));
+						}
+					}
+					ferets[i] = feret;
+				}
+			});
+		}
+		Multithreader.startAndJoin(threads);
+		return ferets;
+	}
+
+	/**
+	 * Get the list of best-fit ellipsoids for the particle surfaces
+	 * 
+	 * @param surfacePoints
+	 * @return Object[] array containing the list of ellipsoids, each of which is
+	 *         also stored as an Object[] array (see FitEllipsoid.yuryPetrov() for
+	 *         details). Note that an Object[] is also an Object so there is no need
+	 *         to make a 2D array (i.e. Object[][]). However, client code must
+	 *         unwrap the ellipsoid elements into arrays by casting to Object[].
+	 */
+	static Object[] getEllipsoids(final Collection<List<Point3f>> surfacePoints) {
+		return surfacePoints.parallelStream().map(surface -> {
+
+			if (surface == null)
+				return null;
+
+			final Iterator<Point3f> pointIter = surface.iterator();
+			final double[][] coOrdinates = new double[surface.size()][3];
+			int i = 0;
+			while (pointIter.hasNext()) {
+				final Point3f point = pointIter.next();
+				coOrdinates[i][0] = point.x;
+				coOrdinates[i][1] = point.y;
+				coOrdinates[i][2] = point.z;
+				i++;
+			}
+
+			Object[] ellipsoid = null;
+			try {
+				ellipsoid = FitEllipsoid.yuryPetrov(coOrdinates);
+			} catch (final RuntimeException re) {
+
```

Comment:
In general you should catch only the specific exception types a method might throw. In this case `IllegalArgumentException`. Otherwise we won't be aware that the code is doing something it shouldn't - throwing unexpected exceptions. Might be also worth to log something here, i.e. `IJ.log("Ellipsoid fitting failed")`